### PR TITLE
Atomic SQLiteNode

### PR DESCRIPTION
**Pull request by tylerkaraszewski**
Note: This is way smaller than it looks. A try/catch block that changed indentation for all of `SQLiteNode::update()` makes it look way bigger if you don't disable whitespace.

### Details
This updates `SQLiteNode` to have all atomic changes.

The read only public APIs (those marked `const`) are all noted as to whether they can block or not. In general, it is a bad idea to call potentially blocking methods in lots of places. I've verified that none of the read APIs that can block on waiting for writes to complete are called from anywhere but the sync thread (which is fine, this is where most of the writing gets done) or from generating responses to `Status` commands, which don't block real traffic.

The only exception is that `leaderCommandAddress` is called from `SQLiteClusterMessenger` for every escalated command, and so can block if we're in the process of `update`ing or `postPoll`ing the sync node. This may be worth fixing or may not, it only happens on followers when they're about to make a network request, and so isn't that scary ([I created an issue for it here](https://github.com/Expensify/Expensify/issues/209160)).

The public write APIs all block (except `broadcast` which is a special case and is documented inline), but they're all seldom used except `postPoll` and `update`. This combined with the above, means that getting the leader address to escalate a command may get delayed by the time it takes for one of these functions to run. This is the only bottleneck I currently see in here that matters.

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/208302
Fixes https://github.com/Expensify/Expensify/issues/208459
Fixes https://github.com/Expensify/Expensify/issues/208438

### Tests
All the existing tests.

Testing against d2d93ffc7265924f7dfe1500e52f8376128d13ed:

Auth:
```
expensidev2004:/vagrant/Auth/test: ./authtest -threads 8
✅ AccountTest::testGetDeletedAccount
✅ AddCommercialFeedTest::testFailureType
... MORE TESTS HERE ...
✅ WalletPaymentTest::testDebitCardPullTransactions
✅ WalletTransferBalanceTest::testBatchReimbursementsAfterWalletBalanceTransferToDebitCard

[ TEST RESULTS ] Passed: 1812, Failed: 0
expensidev2004:/vagrant/Auth/test:

```

ExpensifyBackupMananger:
```
expensidev2004:/vagrant/ExpensifyBackupManager/test: ./expensifyBackupManagerTest
--------------
✅ BackupTest::keyLoading
✅ BackupTest::upload
✅ BackupTest::download
✅ BackupTest::flagFile
--------------
✅ ClusterBackupTest::testFullClusterUp
✅ ClusterBackupTest::testHalfClusterStateCheck

[ TEST RESULTS ] Passed: 6, Failed: 0
```

FuzzyBot:
```
expensidev2004:/vagrant/FuzzyBot/test: ./conciergetest -threads 4
✅ CaptureAgentActivity::testCaptureConciergeActivity
✅ CaptureAgentActivity::testCaptureExpensiworksActivity
... MORE TESTS HERE ...
✅ InsertFragmentsTest::testFragments
✅ ResetPluginTest::test

[ TEST RESULTS ] Passed: 38, Failed: 0
expensidev2004:/vagrant/FuzzyBot/test:
```

### CODE REVIEW

File: sqlitecluster/SQLiteNode.h [hunk 1]
```diff
@@ -92,34 +89,61 @@ class SQLiteNode : public STCPManager {
...
+    // Does not block.
     State leaderState() const;
 
     // Tell the node a commit has been made by another thread, so that we can interrupt our poll loop if we're waiting
     // for data, and send the new commit.
+    // Does not block.
     void notifyCommit() const;
 
+    [[deprecated("Use HTTP escalation")]]
+    bool peekPeerCommand(SQLite& db, SQLiteCommand& command) const;
```
Comment (rafecolton) [inline sqlitecluster/SQLiteNode.h:137]: Why is this still needed / used if HTTP escalation is now the default?

File: sqlitecluster/SQLiteNode.cpp [hunk 2]
```diff
@@ -138,36 +138,29 @@ SQLiteNode::~SQLiteNode() {
...
         delete socket;
     }
-    _acceptedSocketList.clear();
+    _unauthenticatedIncomingSockets.clear();
 
     for (SQLitePeer* peer : _peerList) {
-        // Shut down the peer
-        if (peer->socket) {
-            _socketList.remove(peer->socket);
-        }
```
Comment (rafecolton) [inline sqlitecluster/SQLiteNode.cpp:150]: Why is this no longer needed?

File: sqlitecluster/SQLiteNode.cpp [hunk 3]
```diff
@@ -326,81 +326,52 @@ bool SQLiteNode::_isNothingBlockingShutdown() const {
...
-        if (_escalatedCommandMap.size()) {
-            SWARN("Abandoned " << _escalatedCommandMap.size() << " escalated commands.");
-            for (auto& commandPair : _escalatedCommandMap) {
-                commandPair.second->response.methodLine = "500 Abandoned";
-                commandPair.second->complete = true;
-                _server.acceptCommand(move(commandPair.second), false);
-            }
-            _escalatedCommandMap.clear();
-        }
-        _changeState(SEARCHING);
```
Comment (rafecolton) [inline sqlitecluster/SQLiteNode.cpp:347]: Do we need to keep this line?

File: sqlitecluster/SQLiteNode.cpp [hunk 4]
```diff
@@ -571,9 +552,11 @@ bool SQLiteNode::update() {
         SASSERTWARN(!_syncPeer);
         SASSERTWARN(!_leadPeer);
         SASSERTWARN(_db.getUncommittedHash().empty());
-        // If we're trying to shut down, just do nothing
-        if (shutdownComplete())
+        // If we're trying to shut down, just do nothing, especially don't jump directly to leading and get stuck in an
+        // endless loop.
+        if (_gracefulShutdown()) {
```
Comment (rafecolton) [inline sqlitecluster/SQLiteNode.cpp:557]: Why change this from `shutdownComplete()` to `_gracefulShutdown()`?

File: sqlitecluster/SQLiteNode.cpp [hunk 5]
```diff
@@ -1174,7 +1155,6 @@ bool SQLiteNode::update() {
         if (STimeNow() > _stateTimeout) {
             // Give up
             SHMMM("Timed out waiting for SUBSCRIPTION_APPROVED, reconnecting to leader and re-SEARCHING.");
-            unique_lock<shared_mutex> leadPeerLock(_stateMutex);
```
Comment (rafecolton) [inline sqlitecluster/SQLiteNode.cpp:1177]: Why remove this?

File: sqlitecluster/SQLiteNode.cpp [hunk 6]
```diff
@@ -1242,583 +1222,606 @@ bool SQLiteNode::update() {
...
-        STHROW("unrecognized message");
+            STHROW("unrecognized message");
+        }
+    } catch (const SException& e) {
+        PWARN("Error processing message '" << message.methodLine << "' (" << e.what() << "), reconnecting.");
+        SData reconnect("RECONNECT");
+        reconnect["Reason"] = e.what();
+        peer->sendMessage(reconnect.serialize());
+        peer->shutdownSocket();
     }
```
Comment (rafecolton) [inline sqlitecluster/SQLiteNode.cpp:1825]: Were we not catching this before or is this necessary because we're doing ping/pong handling directly in this method now?

File: sqlitecluster/SQLiteNode.cpp [hunk 7]
```diff
@@ -2557,11 +2531,12 @@ bool SQLiteNode::hasQuorum() const {
 }
 
 void SQLiteNode::prePoll(fd_map& fdm) const {
```
Comment (rafecolton) [inline sqlitecluster/SQLiteNode.cpp:2533]: Can you please summarize what changes to this method are functional vs refactoring? There's a lot of change so it's a bit hard to follow

**Review by bondydaa — COMMENTED** [review]
couple of comments so far. still have more to get through, i'll do that next week.

File: sqlitecluster/SQLiteNode.h (hunk 1 above)
Comment (tylerkaraszewski) [reply sqlitecluster/SQLiteNode.h:137]: `HTTP escalation` is the default for things that used to use `ESCALATE`. This handles `SYNCHRONIZE` but this could pretty easily be moved to use HTTP escalation in the near future.

File: sqlitecluster/SQLiteNode.cpp (hunk 2 above)
Comment (tylerkaraszewski) [reply sqlitecluster/SQLiteNode.cpp:150]: _socketList is gone.

File: sqlitecluster/SQLiteNode.cpp (hunk 3 above)
Comment (tylerkaraszewski) [reply sqlitecluster/SQLiteNode.cpp:347]: No, because when this returns we fall out of this main loop and this becomes irrelevant:
https://github.com/Expensify/Bedrock/blob/6d84191ca0afdb55ad26f22957cfb91e9a547df1/BedrockServer.cpp#L660

Leaving it in actually caused some weirdness in one of the intermediate commits in this branch though I can't remember exactly what it was now. I did investigate that it was safe to remove.

Comment (tylerkaraszewski) [reply sqlitecluster/SQLiteNode.cpp:347]: Figured out what the weirdness was: We made this method `const` so it can no longer change state, but this way we can shared_lock it instead of unique_lock.

And because as soon as we would have made this change and then returned `true`, it becomes irrelevant anyway, the node is getting cleaned up and destroyed.

File: sqlitecluster/SQLiteNode.cpp (hunk 4 above)
Comment (tylerkaraszewski) [reply sqlitecluster/SQLiteNode.cpp:557]: `shutdownComplete()` is public and thus locks, and our convention for this file is that public const functions lock_shared, public non-const functions lock_unique, and private functions do not lock (and thus can't call public functions).

File: sqlitecluster/SQLiteNode.cpp (hunk 5 above)
Comment (tylerkaraszewski) [reply sqlitecluster/SQLiteNode.cpp:1177]: We lock once at the sop of `update` now: https://github.com/Expensify/Bedrock/pull/1252/files#diff-2d1c5f24a4b7ebc3076630d8ee3a1155ad5e7789abe72cff977e615541c319d6R536

File: sqlitecluster/SQLiteNode.cpp (hunk 7 above)
Comment (tylerkaraszewski) [reply sqlitecluster/SQLiteNode.cpp:2533]: It should have the same effect as before, but no longer needs direct access to the peer's socket, which is now private.

File: sqlitecluster/SQLiteNode.cpp (hunk 6 above)
Comment (tylerkaraszewski) [reply sqlitecluster/SQLiteNode.cpp:1825]: We essentially moved the `try/catch` block inside `_onMESSAGE` from here: https://github.com/Expensify/Bedrock/pull/1252/files#diff-2d1c5f24a4b7ebc3076630d8ee3a1155ad5e7789abe72cff977e615541c319d6L2788-R2664

**Review by rafecolton — APPROVED** [review]
Not 100% confident I caught every possible bug, but everything here seems to make sense and all of my questions have been addressed 👍🏼
